src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
+
+ @since 2.1.0
 -}
-class FromPFunN (a :: S -> Type) (b :: S -> Type) (c :: PTermType) where
-    fromPFun' ::
-        Proxy c ->
-        (forall s. Term s (a :--> b)) ->
-        TestableFun c (a :--> b)
+class PUnLam' (fin :: S -> Type) (p :: S -> Type) (end :: Bool) where
+    pUnLam' :: (forall s. Term s p) -> PUnLamHask fin p
 
--- | @since 2.0.0
 instance
-    forall (a :: S -> Type) (b :: S -> Type).
-    ( PLift a
-    , PLift b
+    forall (fin :: S -> Type) (pa :: S -> Type) (pb :: S -> Type).
+    ( PUnLam' fin pb (IsFinal fin pb)
+    , PUnLamHask fin (pa :--> pb) ~ (TestableTerm pa -> PUnLamHask fin pb)
     ) =>
-    FromPFunN (a :--> b) PBool 'LastPFunction
+    PUnLam' fin (pa :--> pb) 'False
     where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) = TestableTerm $ f # x
+    pUnLam' f = \(TestableTerm y) -> pUnLam' @fin @pb @(IsFinal fin pb) (f # y)
 
--- | @since 2.0.0
-instance FromPFunN a PBool 'LastPTerm where
-    fromPFun' _ f (TestableTerm x) = TestableTerm $ f # x
+instance PUnLam' fin fin 'True where
+    pUnLam' x = TestableTerm x
 
--- | @since 2.0.0
-instance
-    forall
-        (a :: S -> Type)
-        (b :: S -> Type)
-        (c :: S -> Type)
-        (d :: S -> Type)
-        (e :: S -> Type).
-    ( c ~ (d :--> e)
-    , PLift a
-    , PLift b
-    , FromPFunN d e (PTT c)
-    ) =>
-    FromPFunN (a :--> b) c 'PFunction
-    where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) =
-        fromPFun' (Proxy @(PTT c)) $ f # x
+{- | Constraint for `PUnLam'` that will give a better type error message.
 
--- | @since 2.0.0
-instance
-    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type) (d :: S -> Type).
-    ( b ~ (c :--> d)
-    , FromPFunN c d (PTT b)
-    ) =>
-    FromPFunN a b 'PTerm
-    where
-    fromPFun' _ f (TestableTerm x) = fromPFun' (Proxy @(PTT b)) $ f # x
+ @since 2.1.0
+-}
+type PUnLam fin p = (PUnLam' fin p (IsFinal fin p), CheckReturn fin p)
+
+{- | Wrapper for @pUnLam'@. Only required the final return type.
+
+ @since 2.1.0
+-}
+punlam' ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam' = pUnLam' @fin @p @(IsFinal fin p)
+
+{- | Wrapper for @pUnLam'@. Same as @punlam'@ but evaluates the given
+     Plutarch function before the conversion. It will throw an error
+     if evaluation failes.
+
+ @since 2.1.0
+-}
+punlam ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam pf = punlam' @fin (loudEval pf)
 
-{- | Converts Plutarch Functions into `Testable` Haskell function of
-  TestableTerms.
+{- | Converts Plutarch Functions into Haskell function of
+     TestableTerms. Then, wrap any instance of TestableTerm that is
+     lambda into `PFun`. Result of this function will be
+     "QuickCheck-compatible" given that all Plutarch type used have
+     `PArbitrary` instance.
```

Comment:
"Converts a Plutarch function into a Haskell function on 'TestableTerm's, then wraps functions into 'PFun' as necessary. The result will be \'Quickcheck-compatible\' if all Plutarch types used have 'PArbitrary' instances."